src/day8.scala

```diff
@@ -0,0 +1,130 @@
+// using scala 3.0.2
+
+package day8
+
+import scala.util.Using
+import scala.io.Source
+
+import Segment.*
+import Digit.*
+
+@main def part1: Unit =
+  println(s"The solution is ${part1(readInput())}")
+
+@main def part2: Unit =
+  println(s"The solution is ${part2(readInput())}")
+
+def readInput(): String =
+  Using.resource(Source.fromFile("input/day8"))(_.mkString)
+
+type Segments = Seq[Segment]
+
+def part1(input: String) =
+
+  def getDisplay(line: String) = line.split("[|]")(1).trim
+
+  val uniqueDigits =
+    for
+      display <- input.linesIterator.map(getDisplay)
+      segments <- display.split(" ")
+      uniqueDigit <- parseUniqueDigit(segments)
+    yield
+      uniqueDigit
+
+  uniqueDigits.length
+end part1
+
+def part2(input: String) =
+
+  def parseSegments(segements: String): Seq[Segments] =
+    segements.split(" ").toSeq.map(Segment.parseSeq)
+
+  def splitParts(line: String) =
+    val Array(cypher, plaintext) = line.split("[|]").map(_.trim).map(parseSegments)
+    (cypher, plaintext)
+
+  def digitsToInt(digits: Seq[Digit]) =
+    digits.foldLeft(0)((acc, d) => acc * 10 + d.ordinal)
+
+  val problems = input.linesIterator.map(splitParts)
+
+  val solutions = problems.map((cypher, plaintext) =>
+    plaintext.map(substitutions(cypher))
+  )
+
+  solutions.map(digitsToInt).sum
+
+end part2
+
+val parseUniqueDigit = Segment.parseSeq andThen Digit.lookupUnique
+
+def substitutions(cypher: Seq[Segments]): Segments => Digit =
+
+  def lookup(section: Seq[Segments], withSegments: Set[Segment]): (Segments, Seq[Segments]) =
+    val found = section.find(s => (s.toSet & withSegments) == withSegments).get
```

Comment:
If we transform the segments to a set, do we want to model them as a Seq in the first place? (ie, is order important? uniqueness? I didn’t read the problem statement)

Reply:
Also, it is not clear why  `.get` is always expected to succeed, maybe there are preconditions for `withSegments` that would be worth indicating?

Reply:
> If we transform the segments to a set, do we really want to model them as a Seq in the first place? (ie, is order important? uniqueness? I didn’t read the problem statement)

I agree that it would be nicer to model them as a `Set` from the start.

Reply:
I had some trouble when I was implementing when I tried to use set directly, but now I was able to refactor to only use set, will push soon

Reply:
`lookup` expects that `withSegments` should be a unique subset of one of the elements of `section`, this works with the logic for this specific problem